pkg/controller/endpoint/trigger_time_tracker.go

```diff
@@ -0,0 +1,219 @@
+			// the last sync. Later we would observe the T0 and T1 trigger change times and because both
+			// T0<T2 and T1 < T2 we would export the metric twice.
+			this.lastSyncMinTriggerTime[key] = triggerTime
+		}
+		return
+	}
+
+	// If we are here it means that the triggerTime is after lastSyncMaxTriggerTime. If this is first
+	// such triggerTime per batch, i.e. minTriggerTime is not set, then we should set it. If it was
+	// already set then we don't need to set it anymore for this batch because events are in the
+	// chronological order.
+	if _, ok := this.minTriggerTime[key]; !ok {
+		this.minTriggerTime[key] = triggerTime
+	}
+}
+
+// Method to be called directly before the call to the "list" function.
+func (this *triggerTimeTracker) startListing(key string) {
+	this.mutex.Lock()
+	defer this.mutex.Unlock()
+	this.isListingObjects[key] = true
+}
+
+// Method that should be called directly after the listing objects has ended. It will reset the
+// state for the given endpoints key and return the time that should be exported as the
+// EndpointsLastChangeTriggerTime annotation. It may happen that the method will return nil, in such
+// case the annotation shouldn't be exported.
+func (this *triggerTimeTracker) stopListingAndReset(
+		key string, triggerTimesFromListing []time.Time) (*time.Time){
+	this.mutex.Lock()
+	defer this.mutex.Unlock()
+
+	var endpointsLastChangeTriggerTime time.Time
+
+	if _, ok := this.minTriggerTime[key]; !ok {
+		// There was no event observed that set the minTriggerTime.
+
+		minTriggerTime := func() *time.Time {
+			// Try in the dirty times.
+			if p := minGreaterThan(this.dirtyTriggerTimes[key], this.lastSyncMaxTriggerTime[key]);
+					p != nil {
+				return p
+			}
+			// If nothing found in the dirty times, try in the times from listing objects.
+			if p := minGreaterThan(triggerTimesFromListing, this.lastSyncMaxTriggerTime[key]); p != nil {
+				return p
+			}
+
+			return nil
+		}()
+
+		if minTriggerTime == nil {
+			// Nothing was observed, nothing in the dirty set, nothing fresh enough in listing times.
+			// Reset state and return nil, there is nothing to export this time.
+			this.isListingObjects[key] = false
+			this.dirtyTriggerTimes[key] = nil
+			return nil
+		}
+
+		this.minTriggerTime[key] = *minTriggerTime
+	}
+
+	endpointsLastChangeTriggerTime = this.minTriggerTime[key]
+	delete(this.minTriggerTime, key)
+
+	maxTimeFromListing := max(triggerTimesFromListing)
+	// See if there is anything in dirty times that could have been a potential new minTriggerTime.
+	if p := minGreaterThan(this.dirtyTriggerTimes[key], maxTimeFromListing); p != nil {
+		this.minTriggerTime[key] = *p
+	}
+	// Clear dirty times.
+	this.isListingObjects[key] = false
+	this.dirtyTriggerTimes[key] = nil
+
+	// Updated the lastSync min and max.
+	this.lastSyncMinTriggerTime[key] = endpointsLastChangeTriggerTime
+	this.lastSyncMaxTriggerTime[key] = maxTimeFromListing
```

Comment:
Oh, if I saw this comment earlier I wouldn't be addressing all comments above :)

The lastTransitionTimes coming in the chronological order is one of the most crucial assumptions here. I the assumption doesn't work then we definitely need to try something else.

But just to clarify, are you proposing to rewrite endpoints_controller in the way the proxier.go works? That would eliminate the "two sources of truth" (events and listing from cache) problem that makes this trigger time tracking so complicated. The only disadvantage is that we would significantly change the way the endpoint controller works, which can have latency and/or memory related implications (in contrast here we're only adding something to the existing framework, when it comes to updating the endpoints objects the logic stays exactly the same).
I thought we wanted to avoid that, but if that's on the table then I think it would be the best option here. 

I totally agree that it would simplify trigger time tracking significantly (it would be similar to https://github.com/kubernetes/kubernetes/pull/71999 which is way simpler), but I'm not sure about the "relatively small changes in the endpoints controller". 
Let's sync in person on this. I'm OOO tomorrow, but will be in Poland next week. I'll set up some meeting.